src/Parser.jl

```diff
@@ -48,267 +155,288 @@ function _count_before{T<:AbstractString}(haystack::T, needle::Char, _end::Int)
     return count
 end
 
+
 # Prints an error message with an indicator to the source
-function _error(message::AbstractString, ps::ParserState)
-    lines = _count_before(ps.str, '\n', ps.s)
+function _error(message::AbstractString, ps::MemoryParserState)
+    orig = Compat.UTF8String(ps.utf8data)
+    lines = _count_before(orig, '\n', ps.s)
     # Replace all special multi-line/multi-space characters with a space.
-    strnl = replace(ps.str, r"[\b\f\n\r\t\s]", " ")
+    strnl = replace(orig, r"[\b\f\n\r\t\s]", " ")
     li = (ps.s > 20) ? ps.s - 9 : 1 # Left index
-    ri = min(ps.e, ps.s + 20)       # Right index
+    ri = min(endof(orig), ps.s + 20)       # Right index
     error(message *
       "\nLine: " * string(lines) *
       "\nAround: ..." * strnl[li:ri] * "..." *
       "\n           " * (" " ^ (ps.s - li)) * "^\n"
     )
 end
 
+function _error(message::AbstractString, ps::StreamingParserState)
+    error("$message\n ...when parsing byte '$(current(ps))'")
+end
+
 # PARSING
 
-function parse_array{T<:AbstractString}(ps::ParserState{T}, dictT::Type)
-    incr(ps) # Skip over the '['
-    _array = Any[]
-    chomp_space(ps)
-    charat(ps)==']' && (incr(ps); return _array) # Check for empty array
-    while true # Extract values from array
-        v = parse_value(ps, dictT) # Extract value
-        push!(_array, v)
-        # Eat up trailing whitespace
-        chomp_space(ps)
-        c = charat(ps)
-        if c == ','
-            incr(ps)
-            continue
-        elseif c == ']'
-            incr(ps)
-            break
-        else
-            _error("Unexpected char: " * string(c), ps)
-        end
+"""
+Given a `ParserState` whose current position is after the last delimiter, return
+the next item in the array or object literal.
+"""
+function parse_value(ps::ParserState, dictT::Type)
+    chomp_space!(ps)
+
+    byte = byteat(ps)
+    if byte == STRING_DELIM
+        parse_string(ps)
+    elseif byte == OBJECT_BEGIN
+        parse_object(ps, dictT)
+    elseif DIGIT_ZERO ≤ byte ≤ DIGIT_NINE || byte == MINUS_SIGN
+        parse_number(ps)
+    elseif byte == ARRAY_BEGIN
+        parse_array(ps, dictT)
+    else
+        parse_simple(ps)
     end
-    return _array
 end
 
-function parse_object{T<:AbstractString}(ps::ParserState{T}, dictT::Type)
-    # Handle case where type is not parameterized yet with key type
-    keyT = dictT.parameters[1]
-    if typeof(keyT) == TypeVar
-        parse_object(ps, dictT, AbstractString, dictT{AbstractString, Any}())
-    else
-        parse_object(ps, dictT, keyT, dictT())
+
+@inbounds function parse_array(ps::ParserState, dictT::Type)
+    result = Any[]
+    incr!(ps)  # Skip over opening '['
+    chomp_space!(ps)
+    if byteat(ps) ≠ ARRAY_END  # special case for empty array
+        while true
+            push!(result, parse_value(ps, dictT))
+            chomp_space!(ps)
+            c = byteat(ps)
+            c == ARRAY_END && break
```

Comment:
Since `c` is never used elsewhere, these two lines can be combined into `byteat(ps) == ARRAY_END && break`